flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java

```diff
@@ -276,6 +277,57 @@ public CompletableFuture<YarnWorkerNode> requestResource(
+
+            requestResourceFuture.whenComplete(
+                    (ignore, t) -> {
+                        if (t instanceof ActiveResourceManager.RequestCancelledException) {
+                            try {
+                                final Queue<CompletableFuture<YarnWorkerNode>>
+                                        pendingRequestResourceFutures =
+                                                requestResourceFutures.getOrDefault(
+                                                        taskExecutorProcessSpec,
+                                                        new LinkedList<>());
+                                if (pendingRequestResourceFutures.remove(requestResourceFuture)) {
+                                    log.info(
+                                            "cancelling pending request with priority {}, remaining {} pending container requests.",
+                                            priority,
+                                            pendingRequestResourceFutures.size());
+                                    int pendingRequestsSizeBeforeCancel =
+                                            pendingRequestResourceFutures.size() + 1;
+                                    final Iterator<AMRMClient.ContainerRequest>
+                                            pendingContainerRequestIterator =
+                                                    getPendingRequestsAndCheckConsistency(
+                                                                    priority,
+                                                                    resource,
+                                                                    pendingRequestsSizeBeforeCancel)
+                                                            .iterator();
+
+                                    Preconditions.checkState(
+                                            pendingContainerRequestIterator.hasNext());
+
+                                    final AMRMClient.ContainerRequest pendingRequest =
+                                            pendingContainerRequestIterator.next();
+                                    removeContainerRequest(pendingRequest);
+
+                                    if (pendingRequestResourceFutures.isEmpty()) {
+                                        requestResourceFutures.remove(taskExecutorProcessSpec);
+                                    }
+
+                                    if (getNumRequestedNotAllocatedWorkers() <= 0) {
+                                        resourceManagerClient.setHeartbeatInterval(
+                                                yarnHeartbeatIntervalMillis);
+                                    }
+                                } else {
+                                    log.info(
+                                            "pending request with priority {} canceled, but it is already allocated.",
+                                            priority);
```

Comment:
How could the request already be allocated? Shouldn't access to the request futures always in main thread?